File: examples/node/webhook_receiver.js
// Kollat B2B API — Webhook receiver (Node.js / Express)
// Requires: npm install express

const crypto = require("crypto");
const express = require("express");

const app = express();
const WEBHOOK_SECRET = "your_webhook_secret_here";

// Use raw body for signature verification
app.use("/kollat/events", express.raw({ type: "application/json" }));

function verifySignature(payload, header) {
  if (!header) return false;
  const expected =
    "sha256=" +
    crypto.createHmac("sha256", WEBHOOK_SECRET).update(payload).digest("hex");
  return crypto.timingSafeEqual(Buffer.from(expected), Buffer.from(header));
}

app.post("/kollat/events", (req, res) => {
  const signature = req.headers["x-kollat-signature"] || "";

  if (!verifySignature(req.body, signature)) {
    return res.status(401).json({ error: "Invalid signature" });
  }

  const event = JSON.parse(req.body);
  handleEvent(event);

  res.json({ ok: true });
});

function handleEvent(event) {
  const { type: eventType, data } = event;

  switch (eventType) {
    case "offramp.confirmada":
      console.log(`Payment confirmed: order #${data.orden_id}`);
      break;
    case "offramp.reclamada":
      console.log(`Payment disputed: order #${data.orden_id} — review required`);
      break;
    case "onramp.completada":
      console.log(`On-ramp completed: order #${data.orden_id}`);
      break;
    case "onramp.expirada":
      console.log(`On-ramp expired: order #${data.orden_id}`);
      break;
    default:
      console.log(`Unhandled event: ${eventType}`);
  }
}

app.listen(8080, () => console.log("Webhook receiver running on :8080"));

File: examples/python/webhook_receiver.py
"""
Kollat B2B API — Webhook receiver example (Python / Flask)
Requires: pip install flask
"""

import hashlib
import hmac

from flask import Flask, abort, jsonify, request

app = Flask(__name__)

WEBHOOK_SECRET = "your_webhook_secret_here"


def verify_signature(payload: bytes, header: str) -> bool:
    if not header:
        return False
    expected = "sha256=" + hmac.new(
        WEBHOOK_SECRET.encode(), payload, hashlib.sha256
    ).hexdigest()
    return hmac.compare_digest(expected, header)


@app.route("/kollat/events", methods=["POST"])
def kollat_webhook():
    signature = request.headers.get("X-Kollat-Signature", "")

    if not verify_signature(request.data, signature):
        abort(401, "Invalid signature")

    event = request.json
    handle_event(event)

    return jsonify({"ok": True}), 200


def handle_event(event: dict):
    event_type = event.get("event")
    data = event.get("data", {})

    if event_type == "offramp.confirmada":
        orden_id = data["orden_id"]
        print(f"Payment confirmed: order #{orden_id}")

    elif event_type == "offramp.reclamada":
        orden_id = data["orden_id"]
        print(f"Payment disputed: order #{orden_id} — review required")

    elif event_type == "onramp.completada":
        orden_id = data["orden_id"]
        print(f"On-ramp completed: order #{orden_id}")

    elif event_type == "onramp.expirada":
        orden_id = data["orden_id"]
        print(f"On-ramp expired: order #{orden_id} — recreate if needed")

    else:
        print(f"Unhandled event: {event_type}")


if __name__ == "__main__":
    app.run(port=8080)
